File: 74/Go/main.go
package main

import "fmt"

func BinarySearch(nums []int, target int) bool {
	left, right := 0, len(nums)-1
	for left <= right {
		middle := (left + right) >> 1
		if nums[middle] == target {
			return true
		} else if nums[middle] > target {
			right = middle - 1
		} else {
			left = middle + 1
		}
	}
	return false
}

func searchMatrix(matrix [][]int, target int) bool {
	for i := 0; i < len(matrix)-1; i++ {
		if target >= matrix[i][0] && target < matrix[i+1][0] {
			return BinarySearch(matrix[i], target)
		}
	}
	return BinarySearch(matrix[len(matrix)-1], target)
}
func main() {
	fmt.Println(searchMatrix([][]int{
		{1, 3, 5, 7},
		{10, 11, 16, 20},
		{23, 30, 34, 60},
	}, 3))
	fmt.Println(searchMatrix([][]int{
		{1, 3, 5, 7},
		{10, 11, 16, 20},
		{23, 30, 34, 60},
	}, 13))
}

File: 74/Cpp/main.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        for ( const auto& row : matrix ) {
            auto it = lower_bound(row.begin(), row.end(), target);
            if ( it != row.end() && *it == target ) {
                return true;
            }
        }
        return false;
    }

    // 右边界
    int BinarySearchRight(vector<int>& nums, int target) {
        int left = -1, right = nums.size();
        while (left + 1 < right) {
            int middle = left + ((right - left) >> 1);
            if(nums[middle] > target) {
                right = middle;
            } else {
                left = middle;
            }
        }
        return left;
    }

    // 左边界
    int BinarySearchLeft(vector<int>& nums, int target) {
        int left = -1, right = nums.size();
        while (left + 1 < right) {
            int middle = left + ((right - left) >> 1);
            if (nums[middle] < target) {
                left = middle;
            } else {
                right = middle;
            }
        }
        return right;
    }
};

int main () {
    vector<int> nums = {1, 2, 3, 3, 4, 5};
    Solution s;
    cout << s.BinarySearchLeft(nums, 3) << endl;
    cout << s.BinarySearchRight(nums, 3) << endl;
}